### PR TITLE
feat(BA-4145): Add device partitioning for AUTO_SPLIT + SHARED

### LINKED ISSUE #8427: Implement DevicePartitioner protocol and device-based AUTO_SPLIT allocation
## Motivation

This ticket implements the core device-based partitioning for AUTO_SPLIT mode as specified in BEP-1041. Instead of calculating abstract slot amounts, we now assign specific device IDs to each agent, ensuring device exclusivity. The `DevicePartitioner` protocol provides a clean abstraction for different partitioning strategies (whole devices vs shared resources like memory).

## Required Features

- Add `DevicePartitioner` protocol with:
  - `device_name: DeviceName` property
  - `generate_assignments(devices, agent_ids) -> Mapping[AgentId, Sequence[DeviceId]]` method
- Add `WholeDevicePartitioner` implementation:
  - For CPU cores and accelerators
  - Uses fill-from-front distribution algorithm
  - Ensures device exclusivity (no device assigned to multiple agents)
- Add `SharedDevicePartitioner` implementation:
  - For memory resources
  - All agents share the same device ID
  - Slot amounts split separately
- Add `distribute_devices(device_ids, agent_ids)` function:
  - Implements fill-from-front algorithm: `q, r = divmod(N, M)`
  - First `r` agents get `q + 1` devices, remaining get `q` devices
  - Devices assigned in natural sorted order
- Add `_natural_sort_key(device_id)` function:
  - Handles numeric suffixes correctly (cuda0, cuda1, cuda10 sort properly)
  - Supports complex IDs like nvme0n1p1, nvme0n1p10
- Add `DeviceAssignments` type alias: `Mapping[AgentId, Mapping[DeviceName, Sequence[DeviceId]]]`
- Implement `_generate_auto_split_assignments()` method
- Implement `_apply_device_assignments()` method:
  - Filters each agent's devices to only assigned ones
  - Updates `alloc_map.device_slots` accordingly
  - For memory, recalculates slot amounts based on mode

## Impact

- `src/ai/backend/agent/resources.py` - `ResourceAllocator`, new partitioner classes
- `src/ai/backend/agent/types.py` - New type definitions
- `tests/unit/agent/test_resource_allocation.py` - New and restored tests

## Testing Scenarios

1. **distribute_devices basic**: 5 devices / 3 agents → [2, 2, 1] distribution
1. **distribute_devices edge cases**:
   - 1 device / 3 agents → [1, 0, 0]
   - 10 devices / 2 agents → [5, 5]
   - 14 devices / 5 agents → [3, 3, 3, 3, 2]
1. **Natural sort ordering**:
   - `["cuda0", "cuda1", "cuda10", "cuda2"]` → `["cuda0", "cuda1", "cuda2", "cuda10"]`
   - `["0", "1", "10", "2"]` → `["0", "1", "2", "10"]`
1. **WholeDevicePartitioner**: Verify correct assignment with various device/agent counts
1. **SharedDevicePartitioner**: Verify all agents receive same device ID
1. **AUTO_SPLIT integration**: Multi-agent scenario with mixed device types (CPU, GPU, memory)
1. **Device exclusivity**: Verify no device ID appears in multiple agents' assignments

JIRA Issue: BA-4145

### PR DESCRIPTION
**Pull request by hhoikoo**
resolves #8427 (BA-4145)

## Overview

Implements device-based partitioning for multi-agent AUTO_SPLIT mode, enabling proper resource
isolation where devices (CPU/GPU) are exclusively assigned to agents while memory is shared.

## Problem Statement

- Multi-agent configurations need exclusive device assignment to prevent resource contention
- Previous slot-based approach didn't partition physical devices between agents
- AUTO_SPLIT mode needed device-centric implementation per BEP-1041 design

## Architecture

```mermaid
flowchart TB
    subgraph "Device Discovery"
        LD[_load_resources] --> GDM[GlobalDeviceMap]
        GDM --> GDI[GlobalDeviceInfo per device type]
    end

    subgraph "Partition Types"
        DP[DevicePartition<br/>Exclusive device IDs]
        SP[SlotPartition<br/>Shared slot amounts]
    end

    subgraph "ResourcePartitioner"
        GSA[generate_shared_assignments]
        GASA[generate_autosplit_assignments]
        CDP[_calculate_device_partitions<br/>DiscretePropertyAllocMap]
        CSP[_calculate_slot_partitions<br/>FractionAllocMap]
        GASA --> CDP --> DP
        GASA --> CSP --> SP
    end

    subgraph "Agent Configuration"
        DP --> ADA[_apply_device_assignments]
        SP --> ADA
        ADA --> AC1[Agent1 ComputerContext]
        ADA --> AC2[Agent2 ComputerContext]
    end
```

## Key Components

**ResourcePartitioner**: Static methods for generating device assignments
- `generate_shared_assignments`: All agents see all devices (SHARED mode)
- `generate_autosplit_assignments`: Exclusive partitioning (AUTO_SPLIT mode)

**Partition Types**:
- `DevicePartition`: Exclusive device ID assignment (CPU cores, GPUs)
- `SlotPartition`: Shared slot amount division (memory)

**Partitioning Logic**:
- `DiscretePropertyAllocMap` → `DevicePartition`: Devices distributed via fill-from-front
- `FractionAllocMap` → `SlotPartition`: Slots divided evenly among agents

**distribute_devices Function**: Implements fill-from-front algorithm with natural sorting
- Given N devices and M agents: first r agents get ⌈N/M⌉ devices, rest get ⌊N/M⌋
- Natural sort ensures cuda0, cuda1, cuda2, cuda10 order (not lexicographic)

---

**Checklist:** (if applicable)

- [ ] Milestone metadata specifying the target backport version
- [x] Mention to the original issue
- [ ] Installer updates including:
  - Fixtures for db schema changes
  - New mandatory config options
- [ ] Update of end-to-end CLI integration tests in `ai.backend.test`
- [ ] API server-client counterparts (e.g., manager API -> client SDK)
- [x] Test case(s) to:
  - Demonstrate the difference of before/after
  - Demonstrate the flow of abstract/conceptual models with a concrete implementation
- [ ] Documentation
  - Contents in the `docs` directory
  - docstrings in public interfaces and type annotations

### CODE REVIEW

File: src/ai/backend/agent/resources.py [hunk 1]
```diff
@@ -83,6 +84,148 @@
...
+        agent_ids: Sequence[AgentId],
+    ) -> Mapping[AgentId, Sequence[DeviceId]]:
+        """
+        Assign all devices to all agents (shared access).
+
+        Every agent receives the full list of device IDs. The actual
+        resource amounts are split separately via slot partitioning.
+        """
+        sorted_devices = sorted(device_ids, key=_natural_sort_key)
+        return dict.fromkeys(agent_ids, sorted_devices)
```
[the remark this answers is not in the document]
Comment (hhoikoo) [reply src/ai/backend/agent/resources.py]: Thanks for catching this! Fixed by using a dict comprehension that creates independent lists per agent:
```python
return {agent_id: list(sorted_devices) for agent_id in agent_ids}
```

File: src/ai/backend/agent/resources.py [hunk 2]
```diff
@@ -674,38 +836,183 @@ def _calculate_available_total_slots(self, total_slots: SlotsMap) -> SlotsMap:
...
+                # Fallback: use the original alloc_map (for unknown types)
+                # Log warning to help operators debug unexpected behavior
+                log.warning(
+                    "Unknown AbstractAllocMap type {} for device {}. "
+                    "Device partitioning will not be applied for agent {}.",
+                    type(alloc_map).__name__,
+                    device_name,
+                    agent_id,
+                )
+                filtered_alloc_map = alloc_map
```
Comment (hhoikoo) [inline src/ai/backend/agent/resources.py]: All this can be replaced with having a `.filter_devices()` function defined at AbstractAllocMap that **mutates** the map, to prevent us having to do this branching logic. We are creating a new alloc_map anyways

File: src/ai/backend/agent/resources.py [hunk 3]
```diff
@@ -532,6 +552,135 @@ def device_ids(self) -> Sequence[DeviceId]:
...
+        """
+        Divide slot amounts among agents for shared devices.
+
+        For FractionAllocMap: simple division (fractional amounts allowed)
+        For DiscretePropertyAllocMap: divmod with remainder distribution
+          (first agents get extra slots when total doesn't divide evenly)
+        """
+        num_agents = len(agent_ids)
+        device_slot_names = {slot_name for slot_name, _ in device_info.plugin.slot_types}
+        is_fractional = isinstance(device_info.alloc_map, FractionAllocMap)
```
Comment (jopemachine) [inline src/ai/backend/agent/resources.py]: How about moving this method implementation to `AllocMap`s instead of this type check?

File: src/ai/backend/agent/resources.py [hunk 4]
```diff
@@ -532,6 +552,135 @@ def device_ids(self) -> Sequence[DeviceId]:
...
+class SlotPartition:
+    slots: SlotsMap
+
+
+type Partition = DevicePartition | SlotPartition
+type ResourceAssignments = Mapping[AgentId, Mapping[DeviceName, Partition]]
+
+
+class ResourcePartitioner:
+    _SHARED_DEVICE_NAMES: Final = frozenset({DeviceName("mem")})
```
Comment (jopemachine) [inline src/ai/backend/agent/resources.py]: Is there any reason this constant is bound to the class?

Comment (hhoikoo) [reply src/ai/backend/agent/resources.py]: Good suggestion\! Moved the constant to module level.

File: src/ai/backend/agent/resources.py (hunk 3 above)
Comment (hhoikoo) [reply src/ai/backend/agent/resources.py]: Good call\! Added `AllocationType` enum and `get_allocation_type()` method to `AbstractAllocMap`, implemented in both `DiscretePropertyAllocMap` and `FractionAllocMap`. Now using pattern matching instead of isinstance check.
